design.md

```diff
@@ -0,0 +1,254 @@
+
+## Design Approach
+
+There's two main aspects to my approach I tried to maintain within each package.
+1. Have critical functionality hang off a `Service` type. This allows me to minimize global state, create a recognizable pattern for readers, and utilize interfaces if needed.
+2. Always take the simplest approach. It was made clear in the challenge description a high performing nor a scalable system was being asked for. Therefore, I designed components in a manner I know will not scale in order to keep things simple. More on this in [Performance Concerns](#performance-concerns).
+
+
+## API
+
+Click [here](https://github.com/tjper/teleport/blob/design/proto/jobworker/v1/service_api.proto) for API definition.
+
+## Authentication
+
+Client and server will utilize mTLS.
+
+TLS 1.3 will be used to ensure a secure cipher suite is utilized. As of Go 1.17, the cipher suites are ordered by `crypto/tls` based on range of factors. https://go.dev/blog/tls-cipher-suites
+
+A set of certificates and secrets will exist in the `certs/` directory. All certificates will be signed by the included CA.
+
+## Authorization
+
+Once a client establishes a connection over TLS, the client certificate's `Common Name` will be used to determine which user has connected. A user will only be authorized to interact with jobs they started. All other jobs will be inaccessible to the user's client.
+
+## Critical Libraries
+
+### cgroups
+
+Package cgroups will provide mechanisms to setup, cleanup, and interact with Linux cgroups.
+
+The `Service` type will be responsible for the following:
+- mounting/unmounting cgroup2
+- directory `/cgroup2/jobworker` setup/cleanup
+- creating/removing a cgroup
+- updating cgroup controllers
+- placing a process into a cgroup
+
+#### Types
+
+```
+type Service struct {
+  ID uuid.UUID
+}
+
+func (s Service) Cleanup() error
+func (s Service) CreateCgroup(options ...CgroupOption) (*Cgroup, error)
+func (s Service) RemoveCgroup(id uuid.UUID) error
+func (s Service) PlaceInCgroup(cgroup Cgroup, pid int) error
+
+type Cgroup struct {
+  ID             uuid.UUID
+  Memory         string
+  Cpus           float32
+  DeviceWriteBps string
+  DeviceReadBps  string
+}
+
+type CgroupOption func(*Cgroup)
+
+func WithMemory(limit string) CgroupOption
+func WithCPUs(cpus float32) CgroupOption
+func WithDeviceWriteBps(deviceLimit string) CgroupOption
+func WithDeviceReadBps(deviceLimit string) CgroupOption
+...
+```
+
+---
+
+### job
+
+Package job will provide a `Service` type for managing `Job` instances.
+
+The `Service` type will be responsible for the following.
+- Setup and cleanup of `var/log/jobworker` logging directory.
+- Maintaining a concurrent-safe jobs map with `Job.ID` as the key and `Job` as the value.
+- Provide methods to start, stop, and fetch managed `Job` instances
+
+The `Job` type will be responsible for the following.
+- Provide unexported mechanisms to start and stop arbitrary commands.
```

Comment:
Can you elaborate on the process execution model? How exactly do you plan on starting a command and ensuring it runs in the correct cgroup?

Reply:
I'm realizing how I've designed this is incorrect. I think I know have a grasp on how to approach this, and am going to make some significant changes.

Reply:
Alright I've made the changes, and they're laid out below in Starting and Stopping sections below.